### ISSUE TITLE
VB Failure On Static Homepage

### ISSUE THREAD

**Issue by timayhelen** [elegantthemes/Divi-Beta#445]
I have seen many reports of the VB failing, and below the page is rendered all of the react components. I few have claimed that it only occurs when editing a static homepage, although it could be a coincidence. 

https://www.elegantthemes.com/forum/viewtopic.php?f=187&t=590950

**Comment by SlavaET** [comment]
I got the login details from customers, but I couldn't figure out the root cause so far. Looks like our ETBuilderBackend variable which stores all the data we need to pass into the app is too big for some environments. Or something wrong inside of this var and it breaks the script in Visual Builder.
I asked for FTP details to perform some tests and get access to wp logs, waiting for customer reply. 
It's weird that the issue appears on homepage only though.

**Comment by timayhelen** [comment]
More people with this issue:

https://www.elegantthemes.com/forum/viewtopic.php?f=187&t=593681

**Comment by SlavaET** [comment]
Looking into it. One customer sent me the FTP details, so I'm experimenting on his server. Looks like there is a limit somewhere which do not allow to load the big amount of data we need for our app, looking the ways how to optimize the data size and/or loading of this data. 
It's weird that it happens only on homepage, probably there is a setting to limit max request length on homepage. This issue reminds me the one we encountered when introduced design settings ( in Divi 2.4 I believe ) when people started to get `500 Internal Server Error`.

**Comment by timayhelen** [comment]
@SlavaET It's strange because the cap we were hitting before was a 2.5MB cap on Document Size. Even a large page in the VB is only around 80kb right now, so it must be a different limit we are reaching.

**Comment by timayhelen** [comment]
@SlavaET There is another person in this thread with the "homepage" problem:

https://www.elegantthemes.com/forum/viewtopic.php?f=187&t=589163

**Comment by SlavaET** [comment]
Yes, it's different for sure, but it's definitely some limit on the server, because when I remove some items it loads fine ( these items also loaded if I load only them, so it's definitely related to the amount of data ). I couldn't figure what is it yet.

**Comment by SlavaET** [comment]
I'm talking about $helpers variable which includes all the module fields. This variable passed to the bundle.js script using the wp_localize_script() and it's quite huge. The problematic part of this variable is advanced settings, if I remove the advanced settings from the variable - it starts loading. If I reduce them, then I see that app is breaking in different point, of the script loading. Looks like advanced settings take a half of the overall size of the $helpers variable.

**Comment by lots0logs** [comment]
Could we move the english strings into the bundle and then request the translations from the server when FB starts only if we detect that the current language is not English?

**Comment by timayhelen** [comment]
We could try raising post_max_size when the VB launches, like we do with php memory limit. Can you see if it fixes the issue on the customer's site?

**Comment by SlavaET** [comment]
No, it doesn't seem to work. If it was the case then it will break the Visual Builder on all the pages I assume. But this weird issue appears only on homepage.. investigating further.

**Comment by SlavaET** [comment]
I couldn't find the root cause of the issue, but I think I found how to overcome it and at the same time improve the loading speed of our app.
# The problem ( as far as I can tell right now ):
In some environments huge `<script>` is not loading correctly and causes 500 internal server error, which breaks the Visual Builder:
![image](https://cloud.githubusercontent.com/assets/3446682/18741269/0521069a-80d0-11e6-9139-279a31672405.png)
I checked our `$helpers` variable which we use to send all the required data to our bundle script and this array contains around 58k strings. Almost Half of its size ( about 28k strings ) is Advanced Settings data. If I remove the advanced settings from the `$helpers` variable, then error disappears and everything work ( except for the advanced options obviously ).
I tried to found the root cause, it looks like there is a limit or something. I checked all the server settings I could think of with no luck.
The most weird thing - it happens only on homepage. It doesn't happen on any other page even if I triple size the `$helper` variable.
# The solution I came up with:
Remove the advanced settings from `$helpers` variable and load it via ajax on the app startup and reduced the size of Advanced Settings array by removing unneeded fields from there ( fields with type = 'skip' which are not used in VB ). I made a PR with this solution and tested it on the customer's website - it seems to work and I didn't notice any side effect during my tests. It should even speed up the app loading, since we will load the half of the VB data via ajax, but it's not really noticeable on localhost.
I'll check my solution on some more customer websites as soon as I get the login details, but it works great on one where I investigated this issue.

**Comment by SlavaET** [comment]
I checked on one more website with this issue and looks like removing Advanced Settings from the helpers is not enough there. So I went ahead and updated my fix to load ALL the module fields ( General, Advanced and CSS ) via ajax. It works on all the websites where I tested it.

**Comment by lots0logs** [comment]
@SlavaET What do you think about what I mentioned in my previous comment. We could avoid the extra ajax request altogether for English sites if we include the untranslated strings in the bundle. That way we only make the request when we need the translated strings (for non-english sites).

**Comment by SlavaET** [comment]
Translations do not need extra ajax requests, we send them via localization as before, so I don't think it may help. It'll reduce the size of helpers viable though, but for English sites only.

**Comment by lots0logs** [comment]
The majority of builder users will have English sites :wink:  Anyway, I was thinking that the advanced settings was just a bunch of translatable strings. I should have double checked it first :sweat_smile:

**Comment by SlavaET** [comment]
Also, if we'll store all strings in the bundle it means we'll need to duplicate them and keep in sync in 2 placed, so I'm not sure we need that. Moreover strings are not so huge comparing to objects, so we can keep it as is I think.

**Comment by SlavaET** [comment]
One thing I worry about is that we include the content object into helpers variable as well. It may cause the same error if user has a big page with lots of modules. Probably we should try loading the content object via ajax as well, before starting the application.

**Comment by lots0logs** [comment]
I agree. Though now I'm beginning to worry about the number of ajax requests :grimacing:  Though really its fine since the VB is only accessed by logged in users :+1:

**Comment by SlavaET** [comment]
We retrieve all the module fields in a single ajax request when app is loading. So it's not an issue at all. I'm going to try loading the content via ajax request, so it'll be one more. 3 in total, I don't think it's too much.

**Comment by SlavaET** [comment]
Added loading of the shortcode object via ajax as well. I combined loading of all the data into single Ajax requests, since it's really small, so Divi app will make one small ajax request on startup to retrieve the required data. The size of retrieved data may vary depending on the page layout but it shouldn't be extremely huge, for example for Homepage Extended layout it's 768kb.
Below are some numbers to compare.

The size of `<script>` tag containing all the VB data for Homepage Extended layout:
- master - `1068102` chars. May vary depending on the page layout;
- my PR - `151490` chars. Almost the same for any page regardless the page layout.

App loading speed with Homepage Extended layout:
- master - `4.7 - 5` seconds;
- my PR - `2.7 - 3` seconds.

So, I see only advantages of this approach, but most important thing - it fixes the weird issue when VB is not loading on homepage on some websites :)

@timayhelen @Jronk @uportnykh @lots0logs would really like to hear your opinion on this.
